x-pack/plugins/case/server/client/comments/add.ts

```diff
@@ -10,135 +10,297 @@ import { pipe } from 'fp-ts/lib/pipeable';
     }),
-    userActionService.postUserActions({
-      client: savedObjectsClient,
-      actions: [
-        buildCommentUserActionItem({
-          action: 'create',
-          actionAt: createdDate,
-          actionBy: { username, full_name, email },
-          caseId: myCase.id,
-          commentId: newComment.id,
-          fields: ['comment'],
-          newValue: JSON.stringify(query),
-        }),
-      ],
+    service.getSubCase({
+      client,
+      id,
     }),
   ]);
 
-  return CaseResponseRt.encode(
-    flattenCaseSavedObject({
-      savedObject: {
-        ...myCase,
-        ...updatedCase,
-        attributes: { ...myCase.attributes, ...updatedCase.attributes },
-        version: updatedCase.version ?? myCase.version,
-        references: myCase.references,
-      },
-      comments: comments.saved_objects,
-    })
+  if (subCasePromise.status === 'fulfilled') {
+    if (subCasePromise.value.references.length > 0) {
+      const caseValue = await service.getCase({
+        client,
+        id: subCasePromise.value.references[0].id,
+      });
+      return new CommentableCase({
+        collection: caseValue,
+        subCase: subCasePromise.value,
+        service,
+        soClient: client,
+      });
+    } else {
+      throw Boom.badRequest('Sub case found without reference to collection');
+    }
+  }
+
+  if (casePromise.status === 'rejected') {
+    throw casePromise.reason;
+  } else {
+    return new CommentableCase({ collection: casePromise.value, service, soClient: client });
+  }
+}
+
+interface AddCommentArgs {
+  caseClient: CaseClientImpl;
+  caseId: string;
+  comment: CommentRequest;
+  savedObjectsClient: SavedObjectsClientContract;
+  caseService: CaseServiceSetup;
+  userActionService: CaseUserActionServiceSetup;
+  user: User;
+}
+
+export const addComment = async ({
+  savedObjectsClient,
+  caseService,
+  userActionService,
+  caseClient,
+  caseId,
+  comment,
+  user,
+}: AddCommentArgs): Promise<CollectionWithSubCaseResponse> => {
+  const query = pipe(
+    CommentRequestRt.decode(comment),
+    fold(throwErrors(Boom.badRequest), identity)
   );
+
+  if (isCommentRequestTypeGenAlert(comment)) {
+    return addGeneratedAlerts({
+      caseId,
+      comment,
+      caseClient,
+      savedObjectsClient,
+      userActionService,
+      caseService,
+    });
+  }
+
+  decodeCommentRequest(comment);
+  const createdDate = new Date().toISOString();
+
+  const combinedCase = await getCombinedCase(caseService, savedObjectsClient, caseId);
+
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  const { username, full_name, email } = user;
+  const userInfo: User = {
+    username,
+    full_name,
+    email,
+  };
+
+  const { comment: newComment, commentableCase: updatedCase } = await combinedCase.createComment({
+    createdDate,
+    user: userInfo,
+    commentReq: query,
+  });
+
+  if (newComment.attributes.type === CommentType.alert && updatedCase.settings.syncAlerts) {
```

Comment:
not necessary for this PR, but could be helpful to just have all these checks encapsulated in functions like `isAlert()` `isCollection()` `isGeneratedAlert` etc... especially if the criteria for any of these changes in the future

Reply:
Great idea. I'll add it to the refactoring list 👍